Lab_Management/PowerShell/Quota/Get-LabServicesCapacityQuotas.ps1

```diff
@@ -0,0 +1,104 @@
+
+param
+(
+    [Parameter(Mandatory=$false, HelpMessage="If you would like more data produced into a CSV file, provide the filename.")]
+    [string] $OutputCSV
+)
+
+$subscriptionId = (Get-AzContext).Subscription.Id
+
+$labsSizes = [Ordered]@{
+    Fsv2 = [Ordered]@{
+        "Small" = 2
+        "Medium" = 4
+        "Large" = 8
+        }
+    Dsv4 = [Ordered]@{
+        "Medium (Nested Virtualization)" = 4
+        "Large (Nested Virtualization)" = 8
+        }
+    NCv3T4 = [Ordered]@{
+        "Small GPU (Compute)" = 6
+        }
+    NVv4 = [Ordered]@{
+        "Small GPU (Visualization)" = 8
+        "Medium GPU (Visualization)" = 12
+        }
+}
+
+$authHeader = @{
+    'Content-Type'='application/json'
+    'Authorization'='Bearer ' + $((Get-AzAccessToken).Token)
+}
+
+# Let's figure out all regions where Lab Services is used:
+$restUri = "https://management.azure.com/subscriptions/$subscriptionId/providers/microsoft.labservices?api-version=2014-04-01-preview"
+$response = Invoke-RestMethod -Uri $restUri -Method Get -Headers $authHeader
+
+# List of all regions available for lab services
+$availableRegionsForLabs = ($response.resourceTypes | Where-Object {$_.resourceType -ieq "labplans"}).Locations
+
+# Save a table of region short names to display names
+$locations = Get-AzLocation
+
+$quotas = $availableRegionsForLabs | ForEach-Object {
+    $regionDisplayName = $_
+    # Get the short code for the region:
+    $region = ($locations | Where-Object {$_.DisplayName -ieq $regionDisplayName} | Select -First 1).Location
+    
+    $authHeader = @{
+        'Content-Type'='application/json'
+        'Authorization'='Bearer ' + $((Get-AzAccessToken).Token)
+    }
+
+    $restUri = "https://management.azure.com/subscriptions/$subscriptionId/providers/Microsoft.LabServices/locations/$region/usages?api-version=2021-11-15-preview"
+    try {
+        $response = Invoke-RestMethod -Uri $restUri -Method Get -Headers $authHeader
+
+        $response.value | Where-Object {$_.limit -gt 0} |
+                Select-Object -Property `
+                    @{Name = 'Region'; Expression = {$region}},
+                    @{Name = 'RegionName'; Expression = {$regionDisplayName}},
+                    @{Name = 'Size'; Expression = {$_.name.value}},
+                    @{Name = "LabSize"; Expression = {$labsSizes[$_.name.value].Keys -Join ", "}},
+                    @{Name = 'UsedCores'; Expression = {$_.currentValue}},
+                    @{Name = 'TotalCores'; Expression = {$_.limit}},
+                    @{Name = 'PercentUsed'; Expression = {[Math]::Round($_.currentValue / $_.limit * 100)}},
+                    @{Name = 'AvailableVMs'; Expression = {
+                            $availableCores = $_.limit - $_.currentValue
+                            ($labsSizes[$_.name.value].GetEnumerator() | ForEach-Object {
+                                "$([Math]::Floor($availableCores / $_.Value)) $($_.Name)"
+                            }) -join ', ' }}
```

Comment:
Will using a comma to join all the available sizes interfere with the outputting the data to a CSV file? If it does, perhaps use `' or '` as the string to join the sizes?

Reply:
The Export-CSV command does the right thing here - the output is properly quoted so that I can open it directly in Excel without a problem...   I can switch it to pipes just in case.